src/components/TimeseriesChart/CursorOverview.tsx

```diff
@@ -0,0 +1,178 @@
+  border-radius: 4px;
+  border: 1px solid #888888;
+  box-sizing: border-box;
+`;
+
+const DateContainer = styled.div`
+  position: absolute;
+  top: 0;
+  margin-left: -260px;
+  width: 220px;
+  color: black;
+  background: #fff;
+  padding: 8px 16px 10px;
+  pointer-events: none;
+  opacity: 0.9;
+  border-radius: 4px;
+  border: 1px solid #888888;
+  box-sizing: border-box;
+`;
+
+const Tag = styled.div`
+  color: black;
+  &:before {
+    content: '';
+    display: inline-block;
+    width: 10px;
+    height: 10px;
+    border-radius: 100%;
+    margin: 0 8px 0 2px;
+    background: ${props => props.color};
+  }
+`;
+
+const formattedDate = (timestamp: number) =>
+  moment(timestamp).format('MMM D, YYYY HH:mm:ss');
+
+export interface CursorOverviewStyles {
+  container?: React.CSSProperties;
+}
+interface CursorOverviewState {
+  overviewContainer: HTMLElement | null;
+}
+export interface CursorOverviewProps {
+  styles?: CursorOverviewStyles;
+  legendValues: LegendValues;
+  series: any;
+  hiddenSeries: { [id: number]: boolean };
+  ruler: RulerConfig;
+}
+
+export class CursorOverview extends React.Component<
+  CursorOverviewProps,
+  CursorOverviewState
+> {
+  static defaultProps = {
+    legendValues: {},
+  };
+
+  overviewContainer: HTMLElement | null = null;
+
+  dateContainer: HTMLElement | null = null;
+
+  constructor(props: CursorOverviewProps) {
+    super(props);
+
+    this.state = {
+      overviewContainer: null,
+    };
+  }
+
+  componentDidMount = () => {
+    window.addEventListener('mousemove', this.handleMouseMove);
+  };
+
+  componentWillUnmount = () => {
+    window.removeEventListener('mousemove', this.handleMouseMove);
+  };
+
+  handleMouseMove = (e: { clientX: number; clientY: number }) => {
+    if (this.overviewContainer) {
+      this.overviewContainer.setAttribute(
+        'style',
+        `transform: translate(${e.clientX + 16}px, ${e.clientY -
+          this.overviewContainer.getBoundingClientRect().height / 2}px)`
+      );
+    }
+
+    const linesContainer = document.getElementsByClassName('lines-container');
+    if (this.dateContainer && linesContainer.length !== 0) {
+      const lineChartRect = linesContainer[0].getBoundingClientRect();
+      this.dateContainer.setAttribute(
+        'style',
+        `transform: translate(${e.clientX + 16}px, ${lineChartRect.top +
+          lineChartRect.height -
+          this.dateContainer.clientHeight -
+          16}px)`
+      );
+    }
+  };
+
+  render() {
+    const { series, hiddenSeries, legendValues, ruler } = this.props;
+    if (!series || series.length <= 0 || !Object.keys(legendValues).length) {
+      return null;
+    }
+
+    const newestTimestamp = Object.keys(legendValues).reduce(
+      (acc: any, id: any) => {
+        const legendValue: LegendValue = legendValues[id];
+        return Math.max(legendValue.rawTimestamp, acc);
+      },
+      0
+    );
+
+    const timeLabel =
+      typeof ruler.timeLabel === 'function'
```

Comment:
There is no sense to recheck type here. Possible options – `undefined` or defined function type. So better just check if `ruler.timeLabel` available

Reply:
What if the client is not using TypeScript?

Plain JS won't catch passing something like:

```js
<TimeseriesChart
    timeseriesIds={[123]}
    startTime={Date.now() - 60 * 1000}
    endTime={Date.now()}
    ruler={{
        yLabel: 100,
        timeLabel: 200,
    }}
/>
```

Reply:
we assume they are passing correct type, we can't spend our time on those ugly types checks.